### PR TITLE
Added 7 floating point regression tests 

### PR DESCRIPTION
**Pull request by bornaehsani**
- float_vec_add() --> A[i] + B[i] --> C[i]
- float_vec_mul() --> A[i] * B[i] --> C[i]
- float_vec_div() --> A[i] / B[i] --> C[i]
- float_vec_exp() --> exp(A[i]) --> C[i]
- float_matrix_mul 
- float_matrix_mul_shared_mem
- float_all_ops() --> Includes all combinations of corner cases (inf, -inf, nan, -nan, 1/0, 0+inf, inf * 0, etc.) and performs all operations (add, sub, mul, div, compare, float to in convert) on a vector.

At the moment all tests pass except test_float_vec_exp(). @vb000 perhaps you can take a look at this test case and it's corresponding RISCV code in bsg_manycore in PR # 154 . 
**merge with PR # 154 in bsg_manycore.**

### CODE REVIEW

**Comment by drichmond** [comment]
This looks good to me. @save-buffer Has also been seeing errors, so I'll pass the final say to him in case there's anything missing!

File: regression/cuda/test_float_all_ops.c [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+	for (int i = 0; i < NUM_CORNER_CASES; i ++) { 
+		A_host[NUM_CORNER_CASES * NUM_CORNER_CASES + NUM_CORNER_CASES + i] = hb_mc_generate_float_rand();
+		B_host[NUM_CORNER_CASES * NUM_CORNER_CASES + NUM_CORNER_CASES + i] = corner_cases[i];
+	}
+
+	// Fill the rest of A and B with random floats
+	for (int i = NUM_CORNER_CASES * NUM_CORNER_CASES + 2 * NUM_CORNER_CASES; i < N; i++) { 
+		A_host[i] = hb_mc_generate_float_rand();
+		B_host[i] = hb_mc_generate_float_rand();
+	}
```
Comment (vb000) [inline regression/cuda/test_float_all_ops.c]: Isn't "all ops" test also testing cases individual op tests consider? Any particular reason we need individual op tests?

File: regression/cuda/test_float_all_ops.c [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+                         float *res_add, float *res_sub,
+                         float *res_mul, float *res_div,
+                         float *res_compare, float *res_convert,
+                         int N){
+	for (int i = 0; i < N; i ++) { 
+		res_add[i] = A[i] + B[i];
+		res_sub[i] = A[i] - B[i];
+		res_mul[i] = A[i] * B[i];
+		res_div[i] = A[i] / B[i];
+		res_compare[i] = (A[i] >= B[i] ? 1 : 0);
```
Comment (vb000) [inline regression/cuda/test_float_all_ops.c]: Could this generate all possible instructions (FLT, FLE, FEQ)?

File: regression/cl_manycore_regression.h [hunk 3]
```diff
@@ -81,6 +81,52 @@ int compare_results(int num_fields, const char *desc[], const uint32_t *expected
         }
         return success ? HB_MC_SUCCESS : HB_MC_FAIL;
 }
+
+
+// Maximum tolerable FP error between X86 and RISCV 
+#define MAX_FLOAT_ERROR_TOLERANCE 1e-6
```
[the remark this answers is not in the document]
Comment (bornaehsani) [reply regression/cl_manycore_regression.h]: Not sure if they are, but I'm seeing differences with the results in the multiplication, division, and expf tests. The relative error is in the order of 1e-34 but still there are differences. Michael said the X86 actually uses 80 bits of precision to RISCV's 32 bits, so I guess marginal errors are expected.

File: regression/cuda/test_float_all_ops.c (hunk 1 above)
Comment (bornaehsani) [reply regression/cuda/test_float_all_ops.c]: The "all ops" test also tries a bunch of corner cases like +- inf, +-nan, division by zero, etc. I guess we can reduce the number of test cases in the "all ops" and have heavy testing in the separate operator tests, but in general you are right, the all ops include all other tests (except expf). @drichmond  what do you think are the separate tests redundant?

File: regression/cuda/test_float_all_ops.c (hunk 2 above)
Comment (bornaehsani) [reply regression/cuda/test_float_all_ops.c]: That's actually a good point, thanks! I'm not sure though, is there a way to ensure it generates all these instructions? Or is there a way to call them specifically on the RISCV?

File: regression/cl_manycore_regression.h [hunk 4]
```diff
@@ -81,6 +81,52 @@ int compare_results(int num_fields, const char *desc[], const uint32_t *expected
...
+	data.hb_mc_int = rand() * (((float)rand() / RAND_MAX) > 0.5 ? 1 : -1) ; 
+	return data.hb_mc_float;
+}
+
+// Compares two floating points and returns the relative error
+// If the original number is zero, returns the differnce to avoid div by zero
+// TODO: find a more appropriate solution for a = 0
+static inline float hb_mc_calculate_float_error (float expected, float result) { 
+	// if expected is close enough to zero that the division would return 1
+	if (fabs(expected) < 1e-15)
```
[the remark this answers is not in the document]
Comment (bornaehsani) [reply regression/cl_manycore_regression.h]: Converting to double is actually a good idea, thanks for pointing it out! but I'm not sure how it can get rid of this condition all together, I still have to make sure div by zero doesn't happen in the error calculation stage. Can you please explain more? thanks!

File: regression/cuda/test_float_all_ops.c (hunk 2 above)
Comment (vb000) [reply regression/cuda/test_float_all_ops.c]: Small error on my part, I realize this is host code, this comment actually goes for the kernel code. Nevertheless, seeing the disassembly is the only way I believe, but that's not practical. So, I think, in kernel code, ops could actually be replaced by precise inline assembly functions with instructions we want to exercise. What do you think?

File: regression/cl_manycore_regression.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (vb000) [reply regression/cl_manycore_regression.h]: To reduce the uncertainty of tolerance, we could have tests that execute ops cumulatively. That way, if small mismatches are due to actual bugs, the error gets accumulated in the end.

Or, we could integrate already available tests into out flow for precise verification. I think prof suggested this: http://www.jhauser.us/arithmetic/TestFloat.html

[the remark this answers is not in the document]
Comment (tommydcjung) [reply regression/cl_manycore_regression.h]: @bornaehsani maybe you can provide a specific example where the multiplication and addition result differ (in hexadecimal)?

[the remark this answers is not in the document]
Comment (tommydcjung) [reply regression/cl_manycore_regression.h]: also, bsg_fpu does not support denormal number as input or output, so make sure that the test is not hitting those case.

File: regression/cl_manycore_regression.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (save-buffer) [reply regression/cl_manycore_regression.h]: Ah, I think I misunderstood the purpose of the if statement. For some reason I'd thought that it was to avoid dividing by too small of a number. I think this should be fine then

**Comment by drichmond** [comment]
Is this ready?

Also - please add license statements to the top if you haven't already

**Comment by bornaehsani** [comment]
@drichmond Added two new regression tests, floating point sqrt and log. With the new newlib patch the expf test passes as well. Added license to all files. 
All tests passed. ready to merge. Merge with PR # 154 in bsg_manycore.
